File: sdk/cpp/tests/test_format.cpp
#include "test_framework.hpp"
#include <anyvali/anyvali.hpp>

using namespace anyvali;
using json = nlohmann::json;

TEST("format: email valid") {
    auto s = string_();
    s->format("email");
    ASSERT(s->safe_parse(json("user@example.com")).success);
}

TEST("format: email missing @") {
    auto s = string_();
    s->format("email");
    auto result = s->safe_parse(json("not-an-email"));
    ASSERT(!result.success);
    ASSERT(result.issues[0].code == "invalid_string");
}

TEST("format: email no dot after @") {
    auto s = string_();
    s->format("email");
    ASSERT(!s->safe_parse(json("user@localhost")).success);
}

TEST("format: url https") {
    auto s = string_();
    s->format("url");
    ASSERT(s->safe_parse(json("https://example.com")).success);
}

TEST("format: url http with path") {
    auto s = string_();
    s->format("url");
    ASSERT(s->safe_parse(json("http://example.com/path?q=1")).success);
}

TEST("format: url rejects ftp") {
    auto s = string_();
    s->format("url");
    ASSERT(!s->safe_parse(json("ftp://files.example.com")).success);
}

TEST("format: uuid valid") {
    auto s = string_();
    s->format("uuid");
    ASSERT(s->safe_parse(json("550e8400-e29b-41d4-a716-446655440000")).success);
}

TEST("format: uuid invalid") {
    auto s = string_();
    s->format("uuid");
    ASSERT(!s->safe_parse(json("not-a-uuid")).success);
}

TEST("format: ipv4 valid") {
    auto s = string_();
    s->format("ipv4");
    ASSERT(s->safe_parse(json("192.168.1.1")).success);
}

TEST("format: ipv4 rejects leading zeros") {
    auto s = string_();
    s->format("ipv4");
    ASSERT(!s->safe_parse(json("192.168.01.1")).success);
}

TEST("format: ipv4 rejects out of range") {
    auto s = string_();
    s->format("ipv4");
    ASSERT(!s->safe_parse(json("256.1.1.1")).success);
}

TEST("format: ipv6 valid full") {
    auto s = string_();
    s->format("ipv6");
    ASSERT(s->safe_parse(json("2001:0db8:85a3:0000:0000:8a2e:0370:7334")).success);
}

TEST("format: ipv6 compressed loopback") {
    auto s = string_();
    s->format("ipv6");
    ASSERT(s->safe_parse(json("::1")).success);
}

TEST("format: ipv6 invalid") {
    auto s = string_();
    s->format("ipv6");
    ASSERT(!s->safe_parse(json("not:an:ipv6")).success);
}

TEST("format: date valid") {
    auto s = string_();
    s->format("date");
    ASSERT(s->safe_parse(json("2024-02-29")).success);
}

TEST("format: date invalid leap day") {
    auto s = string_();
    s->format("date");
    ASSERT(!s->safe_parse(json("2023-02-29")).success);
}

TEST("format: date-time with Z") {
    auto s = string_();
    s->format("date-time");
    ASSERT(s->safe_parse(json("2024-01-15T10:30:00Z")).success);
}

TEST("format: date-time with offset") {
    auto s = string_();
    s->format("date-time");
    ASSERT(s->safe_parse(json("2024-01-15T10:30:00+05:30")).success);
}

TEST("format: date-time without timezone rejected") {
    auto s = string_();
    s->format("date-time");
    ASSERT(!s->safe_parse(json("2024-01-15T10:30:00")).success);
}

File: sdk/ruby/test/test_format.rb
# frozen_string_literal: true

require_relative "test_helper"

class TestFormat < Minitest::Test
  def test_email_valid
    s = AnyVali.string.format("email")
    assert_equal "user@example.com", s.parse("user@example.com")
  end

  def test_email_invalid_no_at
    s = AnyVali.string.format("email")
    result = s.safe_parse("not-an-email")
    assert result.failure?
    assert_equal "invalid_string", result.issues.first.code
    assert_equal "email", result.issues.first.expected
  end

  def test_email_invalid_no_dot
    s = AnyVali.string.format("email")
    result = s.safe_parse("user@localhost")
    assert result.failure?
  end

  def test_url_valid_https
    s = AnyVali.string.format("url")
    assert_equal "https://example.com", s.parse("https://example.com")
  end

  def test_url_valid_http
    s = AnyVali.string.format("url")
    assert_equal "http://example.com/path?q=1", s.parse("http://example.com/path?q=1")
  end

  def test_url_rejects_ftp
    s = AnyVali.string.format("url")
    result = s.safe_parse("ftp://files.example.com")
    assert result.failure?
    assert_equal "invalid_string", result.issues.first.code
  end

  def test_uuid_valid
    s = AnyVali.string.format("uuid")
    assert_equal "550e8400-e29b-41d4-a716-446655440000", s.parse("550e8400-e29b-41d4-a716-446655440000")
  end

  def test_uuid_invalid
    s = AnyVali.string.format("uuid")
    result = s.safe_parse("not-a-uuid")
    assert result.failure?
  end

  def test_ipv4_valid
    s = AnyVali.string.format("ipv4")
    assert_equal "192.168.1.1", s.parse("192.168.1.1")
  end

  def test_ipv4_rejects_leading_zeros
    s = AnyVali.string.format("ipv4")
    result = s.safe_parse("192.168.01.1")
    assert result.failure?
  end

  def test_ipv4_rejects_out_of_range
    s = AnyVali.string.format("ipv4")
    result = s.safe_parse("256.1.1.1")
    assert result.failure?
  end

  def test_ipv6_valid_full
    s = AnyVali.string.format("ipv6")
    assert_equal "2001:0db8:85a3:0000:0000:8a2e:0370:7334", s.parse("2001:0db8:85a3:0000:0000:8a2e:0370:7334")
  end

  def test_ipv6_valid_compressed
    s = AnyVali.string.format("ipv6")
    assert_equal "::1", s.parse("::1")
  end

  def test_ipv6_invalid
    s = AnyVali.string.format("ipv6")
    result = s.safe_parse("not:an:ipv6")
    assert result.failure?
  end

  def test_date_valid
    s = AnyVali.string.format("date")
    assert_equal "2024-02-29", s.parse("2024-02-29")
  end

  def test_date_invalid_leap
    s = AnyVali.string.format("date")
    result = s.safe_parse("2023-02-29")
    assert result.failure?
  end

  def test_datetime_valid_z
    s = AnyVali.string.format("date-time")
    assert_equal "2024-01-15T10:30:00Z", s.parse("2024-01-15T10:30:00Z")
  end

  def test_datetime_valid_offset
    s = AnyVali.string.format("date-time")
    assert_equal "2024-01-15T10:30:00+05:30", s.parse("2024-01-15T10:30:00+05:30")
  end

  def test_datetime_rejects_no_timezone
    s = AnyVali.string.format("date-time")
    result = s.safe_parse("2024-01-15T10:30:00")
    assert result.failure?
  end
end
